forge/test/mlir/llama/test_llama_inference.py

```diff
@@ -126,3 +127,293 @@ def test_llama_inference_cache_cpu():
+    max_new_tokens = padding_seq_len
+    for idx in range(max_new_tokens):
+
+        model_inputs = [input_ids, attention_mask]
+
+        # CPU Inference
+        framework_output = framework_model(input_ids=input_ids, attention_mask=attention_mask)
+
+        # Run on TT device
+        tt_output = compiled_model(*model_inputs)
+        tt_output = [tt_out.to("cpu") for tt_out in tt_output]
+
+        # Validate TT result with Framework
+        assert all([compare_with_golden_pcc(golden=fw_out, calculated=tt_out, pcc=0.99) for fw_out, tt_out in zip(framework_output, tt_output)])
+
+        next_token_index = non_padding_seq_len  + idx
+        next_token_logits = tt_output[0][:, next_token_index - 1, :]
+        next_token = torch.argmax(next_token_logits, dim=-1)
+
+        if next_token == tokenizer.eos_token_id:
+            break
+
+        input_ids[:, next_token_index] = next_token
+        attention_mask[:, next_token_index] = 1
+
+
+    # Generated text
+    generated_text = tokenizer.decode(input_ids[0], skip_special_tokens=True)
+    print("generated_text=",generated_text)
+
+
+@pytest.mark.xfail()
+def test_llama_prefill_on_cpu_decode_on_tt_cache():
+
+    # Load Llama 3B model and tokenizer
+    model_path = "openlm-research/open_llama_3b"
+    framework_model, tokenizer = load_model_and_tokenizer(model_path=model_path, use_cache=True, return_dict=True)
+
+    class LlamaModelWrapper(torch.nn.Module):
+        """
+        In LlamaModelWrapper class, forward function takes single input token (i.e last predicted token)
+        and past key values from the previous iteration and return logits and past key values as List of torch.Tensor
+        Args:
+            input_id (`torch.Tensor`) - shape of (batch_size, 1)
+            past_key_values (`List[List[torch.Tensor, torch.Tensor]]`) - key/values shape - (batch_size, num_of_key_values_heads, key_value_seq_len, head_dim)
+        Returns:
+            outputs (`List[torch.Tensor]`) - Logits of shape (batch_size, 1, vocab_size) will be present in the 0th index of the outputs and
+            rest of the list contains past key values tensor of shape (batch_size, num_of_key_values_heads, key_values_seq_len + 1, head_dim)
+        """
+        def __init__(self, model):
+            super().__init__()
+            self.model = model
+
+        def forward(self, input_id, past_key_values):
+            model_outputs = self.model(input_ids=input_id, past_key_values=past_key_values)
+
+            # TT will return the logits and past key values as list of tensor, so flattening
+            # the past key values from List(List(Key1, Values1), ... , List(Key26, Values26)) to
+            # List(Key1, Values1, ... , Key26, Values26) for comparing the Framework and TT output in similar fashion.
+            outputs = [model_outputs.logits]
+            for k, v in model_outputs.past_key_values:
```

Comment:
Will this post-processing be included in the compiled model?

If that is the case, we should we do this post-processing outside of this wrapper?

Reply:
Sure Nikola. I will move the post processing outside the wrapper.

Reply:
Moved the  post-processing outside the wrapper.